mllib/src/main/scala/org/apache/spark/ml/feature/LSH.scala

```diff
@@ -137,14 +139,23 @@ private[ml] abstract class LSHModel[T <: LSHModel[T]]
       val hashDistUDF = udf((x: Seq[Vector]) => hashDistance(x, keyHash), DataTypes.DoubleType)
       val hashDistCol = hashDistUDF(col($(outputCol)))
 
-      // Compute threshold to get exact k elements.
-      // TODO: SPARK-18409: Use approxQuantile to get the threshold
-      val modelDatasetSortedByHash = modelDataset.sort(hashDistCol).limit(numNearestNeighbors)
-      val thresholdDataset = modelDatasetSortedByHash.select(max(hashDistCol))
-      val hashThreshold = thresholdDataset.take(1).head.getDouble(0)
-
-      // Filter the dataset where the hash value is less than the threshold.
-      modelDataset.filter(hashDistCol <= hashThreshold)
+      val modelDatasetWithDist = modelDataset.withColumn(distCol, hashDistCol)
+      var filtered: DataFrame = null
+      var requestedNum = numNearestNeighbors
+      do {
+        requestedNum *= 2
+        if (requestedNum > modelDataset.count()) {
+          requestedNum = modelDataset.count().toInt
+        }
+        var quantile = requestedNum.toDouble / modelDataset.count()
+        var hashThreshold = modelDatasetWithDist.stat
+          .approxQuantile(distCol, Array(quantile), 0.001)
+
+        // Filter the dataset where the hash value is less than the threshold.
+        filtered = modelDatasetWithDist.filter(hashDistCol <= hashThreshold(0))
```

Comment:
I don't think this is necessary, just ask for progressive higher quantiles of the original dataset. You don't want to both double the number requested _and_ filter out things that were under the previous quantile, I think? just one, and doubling the number requested is much cheaper.

Reply:
Seems to me that I have to filter out to find out if I can get enough number of the nearest neighbors. If not, I go back to the loop to double the quantile. 

I am debating if I should continue this PR. The purpose of this PR is to improve performance. If the first round of the loop doesn't get enough number of the nearest neighbors and we have to go into the loop multiple times, the performance could be worse than the original code. 

In the doc of ```approxNearestNeighbors```, it says ```Given a large dataset and an item, approximately find at most k items which have the closest distance to the item. ``` If this is true, then I guess we can just  use a quantile that should yield 2x the number of results. If we get less than k elements, that's OK. However, the original implementation returns exact k elements. I am not sure if we can change the original behavior.

Reply:
Nevermind my comment; you're not reusing that filtered subset in the next loop, oops. This is correct.
Yes you have to count the filtered result each time.

Certainly the hope is that the first pass will yield enough neighbors, and the loop is just there as a fallback. I think it can be a win if the relative error tolerance is loose enough that 1 or even 2 approx quantile checks is faster than a full sort, but, I don't know how it plays out at scale.

You raise a good point; the docs below do say 'at most k elements', although the current implementation will return exactly k (assuming there are at least k points in the input). It repeats that twice. Hm. I'd also support just making one pass and picking a larger multiple of the request number of neighbors. But I don't mind the approach you have here (modulo a few optimizations above).

We could also optimize the `requestedNum > modelDataset.count()` case by just returning the whole input in that case rather than continuing with another pass.

I wonder if it's reasonable to construct a simple synthetic large input in a test, and test out whether it seems to be faster at some scale, and how you have to set the relative error to get that speedup, and how likely it is that it needs even a second loop.

Reply:
I think over. Seems to me that it's safe not to have the second loop, and we might not need to double the quantile in the beginning.

If the relative error sets to zero, the exact quantile is computed. Filtering out at this exact quantile gives the exact k items of ```numNearestNeighbors``` we ask for. Of course, this could be very expensive. If users want to trade off accuracy for performance, they can increase the relative error. In this case, using approximate quantile as threshold may return less number of data than ```numNearestNeighbors```, but this is OK since the docs says ```approximately find at most k items```. I tried to do some tests but it seems hard to find a universal relative error that works for all cases. Since the approximate quantile has the range of 

```
 floor((p - err) * N) <= rank(x) <= ceil((p + err) * N)
```
In the extreme unlucky case, the approximate quantile is ```(p - err) * N)``` and the elements that meet the requirement happen to be in the range of ```(p - err) * N  to p * N```. It seems to me the appropriate value of relative error depends on the distribution of the dataset and the users may need to adjust their own relative error depends on their dataset. 

So I guess 1) we don't need the loop 2) expose the relative error to user? 
I am OK with either double the quantile or not double it.

Reply:
I take back my last sentence. We should double the quantile to make it safer.

Reply:
To do it in one pass, we need (p - err) * N >= M, where M is the number of nearest neighbors. We'd chose the quantile p to be some multiple k of the desired quantile M/N. I think that works out to needing k >= err * N / M  + 1.   So maybe it's a matter of fixing some err that gets a good speedup, like 0.1, and then just picking quantile p = err + M / N. Makes sense, really: you have to increase the desired quantile by err.

How about something like that as a one-pass solution? It just also means checking that if p > 1 then just fall back to sort + take.

Reply:
Sounds good.

I did a small test to try to find a good value of relative error to use. I used  existing ```BucketedRandomProjectionLSHSuite``` but made the dataset bigger: one with incremental values and one with random values.
```    
val data1 = {
      for (i <- -200 until 200; j <- -200 until 200) yield Vectors.dense(i*10, j*10)
}
dataset1 = spark.createDataFrame(data1.map(Tuple1.apply)).toDF("keys")
```
```    
val data2 = {
      for (i <- -200 until 200; j <- -200 until 200) yield Vectors.dense(Random.nextInt, Random.nextInt)
}
dataset2 = spark.createDataFrame(data2.map(Tuple1.apply)).toDF("keys")
```
So the dataset count is 160000 and I asked for 40000 nearest neighbors. I tested relative error 0.001, 0.005, 0.01, 0.05, 0.1 and  0.2 but didn't see much change in performance.   Not sure if it is because the dataset is too small. 

**Dataset1 with incremental values**

| relative error | 1st run | 2nd run | 3rd run | 4th run | 5th run | average |
| -----------   | ------- | -------- | -------- | -------- | -------| -------|
| 0.001      | 7.61s | 6.56s | 6.39s | 7.32s | 7.49s | 6.998s | 
| 0.005      | 6.39s | 6.44s | 6.62s | 6.39s | 7.54s | 6.67s |
| 0.01      | 6.56s | 6.38s | 7.34s | 6.58s | 6.68s | 6.71s |
| 0.05      | 6.51s | 6.24s | 7.24s | 6.34s | 6.54s | 6.57s |
| 0.1      | 6.28s | 6.20s | 6.34s | 6.68s | 7.07s | 6.51s |
| 0.2     | 6.39s | 6.21s | 6.25s | 6.22s | 6.30s | 6.27s |

**Dataset2 with random values**

| relative error | 1st run | 2nd run | 3rd run | 4th run | 5th run | average |
| -----------   | ------- | -------- | -------- | -------- | -------| -------|
| 0.001      | 7.66s | 6.77s | 6.75s | 7.78s | 6.64s | 7.11s |
| 0.005      | 6.57s | 6.61s | 6.75s | 7.42s | 6.60s | 6.79s |
| 0.01      | 6.68s | 7.44s | 6.25s | 6.69s | 7.48s | 6.91s |
| 0.05      | 6.59s | 6.54s | 6.75s | 6.62s | 6.63s | 6.62s |
| 0.1      | 7.73s | 6.58s | 6.61s | 6.68s | 6.55s | 6.83s |
| 0.2     | 6.61s | 6.62s | 6.54s | 6.51s | 6.59s | 6.57s |


Seems to me that it may not be good to have a fixed value for relative error. For example, 0.05 might be a good relative error for the case of getting 40000 nearest neighbors from 160000 data, but it's too big for the case of getting 400 nearest neighbors from 160000 data. I guess I will pick 
 err = 0.2 M / N. Since p = err + M / N, we actually have p = 1.2 M/N. Hope this makes sense.

Reply:
BTW I wonder how this compares to sort + take?

I think the only issue with making err a function of M/N is that you won't maybe get much speedup when M/N is small, and that's the common case. But I guess you're saying the speed difference isn't that big. Relative error could be smaller than 0.001 though... like 10 nearest neighbors out of 1M = 0.000001. I wonder if that's notably slower?

The downside to err + M/N is that you filter in a lot more elements, although you subsequently sort and take anyway; they won't come back to the driver.

It is a good question whether it overall speeds things up but I think it will if the LSH has a not-tiny relative error.

Reply:
It is much faster than sort + take. 
So I guess we will pick 0.01 as relative error? Or 0.05?

Reply:
Sure, let's start with 0.05. Doesn't look like much gain after that.